### ISSUE TITLE
Setup Vs H2 Settings

### ISSUE THREAD

**Issue by NiteshKant** [ReactiveSocket/reactivesocket#146]
ReactiveSocket has a single unidirectional setup frame and follows a NACK model instead of an ACK model for the setup "properties". This means that a server that does not support a required setup property should reject setup but if all setup properties are supported then no action is required.

This design has the positive of supporting a "false start" i.e. a client does not have to wait for the setup ack. However, it has a few negatives:

- Inability to indicate lack of support for an optional setup property (eg: resumability)
- Inability for a server to send any setup information (eg: server honor lease)

OTOH, HTTP/2 follows a more flexible [Settings frame design](https://tools.ietf.org/html/rfc7540#section-6.5) which has a few advantages:

- Extensible to add new settings.
- Bi-directional
- Supports ack/nack per setting.
- Although trivial but reduce data exchange by omitting properties that one doesn't care about, eg: mime types, resumability.

Should we be thinking about changing this design in the light of the two negatives above?

**Comment by benjchristensen** [comment]
False start is an important characteristic on mobile networks, so not one I'd support eliminating. With resumability we lose it but gain the avoidance of resending everything so the tradeoff is worth it. When resumability is not being used, I don't judge the tradeoff of losing false start worth it. Even with resumability, false start is used when establishing a new connection, the time when false start is most valuable, such as after a RESUME failure and having a large number of queries to re-establish right behind a SETUP frame being sent. Otherwise a RESUME failure has two roundtrips if we have to wait on an ACK after SETUP. 

Some comments on the two negatives: 

> Inability to indicate lack of support for an optional setup property (eg: resumability)

This should not be something being found in production. This should get learned in development. Servers should not be randomly changing their behavior in production environments. And if they do ... well ... extra round trips, latency, failure recovery etc are what would be expected of a client as it adjusts its approach to different code paths (basically having different client behavior for different types of servers it may expect). 

> Inability for a server to send any setup information (eg: server honor lease)

I am not aware of any behavior we need that requires eliminating false start so we can get a setup ack. Server honoring a lease from the client for example is something [I question even needing](https://github.com/ReactiveSocket/reactivesocket/issues/145#issuecomment-274610284), and in general would not consider high enough priority to eliminate false start. 

Now on to possible changes:

> Supports ack/nack per setting.

This is an interesting possibility. Though still not one that I can imagine using in mobile use cases, so it wouldn't even be per setting, it would be up to a client if it cares to wait, or prefers to use false start. 

In a server-to-server environment I can see false start as not being important as the latency is generally hidden behind a pool of connections. However, it's still not clear in that case where latency doesn't really matter (to setup a new connection) why failure (either on a connection or particular request) can't just be used. What does the ACK allow that won't be learned by just starting to send requests? 

> Extensible to add new settings.

Are you suggesting we need something more that either a) additional flags in the SETUP frame, or b) the SETUP data/metadata allowing the application to pass whatever it wishes?

It seems rare to add new settings to the protocol itself, whereas we've given complete flexibility to the application layer to send data/metadata in the SETUP frame and then await an ack using METADATA_PUSH if it so desires. 


I don't think the need is clear enough to change the SETUP frame, or add an ACK at this time. The protocol supports versioning and we could always add an optional SETUP_ACK in a future version if we find a hard requirement for it.

**Comment by tmontgomery** [comment]
What about using LEASE as an implicit ACK if needed?

**Comment by robertroeser** [comment]
@tmontgomery the java client does this I believe - it waits for a request before the client because available. 

ACK instead of NACK seems like it would be useful if the client would start hammer the server before the server finishes setup. If that was the case though you can just use leases, and not send them until you're ready. I can't really of another reason for an ACK.

**Comment by benjchristensen** [comment]
So does that mean that since we have LEASE when wanted, this is a non-issue?

**Comment by tmontgomery** [comment]
Well, the way it is supposed to work is:

"The client-side Requester that has set the L flag in the SETUP frame MUST wait for the server-side Responder to send a LEASE frame before it can send Requests."

This gives the ability to wait and determine if the SETUP is accepted or rejected.

**Comment by NiteshKant** [comment]
Few clarifications from my side:

- I am not advocating adopting h2 settings design, this was just a strawman for discussion.
- I support false start and we should not give it up. Following a settings design does not preclude false start as acks can be optional per setting.

Here is what led me to this thought exercise:

SETUP is a single bag of settings and hence to a degree expects homogeneous properties on all settings. This means either all setup properties require ack or all don't. Is there a middle path or this is good enough?

>>> What about using LEASE as an implicit ACK if needed?

Personally, I do not like overloading leases with acks as that would mean that any setup that requires acks would also require leases (or send a bogus lease frame). So, I would prefer we have an explicit `SETUP_ACK` if required. I do not think we need that at the moment. What is more important as I explain in [this isssue](https://github.com/ReactiveSocket/reactivesocket/issues/144#issuecomment-274943274) is the concept of a NACK for a property which does not close the connection.

So, I think it is fine if we can selectively reject a property in setup instead of `CONNECTION_ERROR` or `REJECTED_SETUP` if we do not think it is worth the shift towards a settings kind of design.

**Comment by benjchristensen** [comment]
It's still not clear to me what would property we'd add that requires an ACK that isn't handled adequately with CONNECTION_ERROR behavior, or doesn't end up with new frames such as happened with RESUME. 

And if we ever come up with a new feature (such as resume was) that needs an ACK, that new feature can have a new ACK frame added, just like RESUME_OK was added. For example, if we do come up with a flag that is in the SETUP frame that requires an ACK, a new frame would be added to return that ACK, and an implementation using that feature would await that ACK. 

So, since we support versioning to add enhancements over time, and we've already seen how this was added gracefully when we added RESUME/RESUME_OK, I don't see a need to add an ACK based on individual properties.
